src/Silex/Provider/SecurityServiceProvider.php

```diff
@@ -140,14 +141,29 @@ public function register(Container $app)
         };
 
         $app['security.access_manager'] = function ($app) {
-            return new AccessDecisionManager($app['security.voters']);
+            $votersIterator = new ServiceIterator($app, $app['security.voter_services']);
+
+            return new AccessDecisionManager($votersIterator);
+        };
+
+        $app['security.voter_services'] = function () {
+            return [RoleHierarchyVoter::class, AuthenticatedVoter::class];
+        };
+
+        $app[RoleHierarchyVoter::class] = function ($app) {
+            return new RoleHierarchyVoter(new RoleHierarchy($app['security.role_hierarchy']));
+        };
+
+        $app[AuthenticatedVoter::class] = function ($app) {
+            return new AuthenticatedVoter($app['security.trust_resolver']);
         };
 
+        // Unused, kept for backwards-compatibility
+        // Extend security.voter_services instead to prevent circular references
         $app['security.voters'] = function ($app) {
-            return [
-                new RoleHierarchyVoter(new RoleHierarchy($app['security.role_hierarchy'])),
-                new AuthenticatedVoter($app['security.trust_resolver']),
-            ];
+            return array_map(function ($voterServiceId) use ($app) {
+                return $app[$voterServiceId];
+            });
```

Comment:
this is broken. It misses the second argument of `array_map`.

and it is not BC either: extending `security.voters` to register additional voters is not working anymore, which is a BC break. This is what should preserve BC.

Reply:
@stof Thanks for that! Fix on the `array_map` is coming shortly. However, this functionality was not documented before. Should we consider that as a BC break?

Reply:
To preserve backwards compatibility we could check whether the list of voter services is the same as the default one and if it is we could use `security.voters` as before.

Reply:
@stof Addressed in #1619.